core-compose/src/main/java/com/squareup/workflow/ui/compose/WorkflowContainer.kt

```diff
@@ -0,0 +1,323 @@
+ *
+ * @param workflow The [Workflow] to render.
+ * @param props The props to render the root workflow with. If this value changes between calls,
+ * the workflow runtime will re-render with the new props.
+ * @param diagnosticListener A [WorkflowDiagnosticListener] to configure on the runtime.
+ * @param content A [Composable] function that gets executed every time the root workflow spits
+ * out a new rendering.
+ */
+@Direct
+@Composable fun <PropsT, RenderingT> WorkflowContainer(
+  workflow: Workflow<PropsT, Nothing, RenderingT>,
+  props: PropsT,
+  modifier: Modifier = Modifier,
+  diagnosticListener: WorkflowDiagnosticListener? = null,
+  content: @Composable() (rendering: RenderingT) -> Unit
+) {
+  WorkflowContainer(workflow, props, {}, modifier, diagnosticListener, content)
+}
+
+/**
+ * Render a [Workflow]'s renderings.
+ *
+ * When this function is first composed it will start a new runtime. This runtime will be restarted
+ * any time [workflow], [diagnosticListener], or the `CoroutineContext`
+ * changes. The runtime will be cancelled when this function stops composing.
+ *
+ * @param workflow The [Workflow] to render.
+ * @param diagnosticListener A [WorkflowDiagnosticListener] to configure on the runtime.
+ * @param content A [Composable] function that gets executed every time the root workflow spits
+ * out a new rendering.
+ */
+@Direct
+@Composable fun <RenderingT> WorkflowContainer(
+  workflow: Workflow<Unit, Nothing, RenderingT>,
+  modifier: Modifier = Modifier,
+  diagnosticListener: WorkflowDiagnosticListener? = null,
+  content: @Composable() (rendering: RenderingT) -> Unit
+) {
+  WorkflowContainer(workflow, Unit, {}, modifier, diagnosticListener, content)
+}
+
+/**
+ * Render a [Workflow]'s renderings.
+ *
+ * When this function is first composed it will start a new runtime. This runtime will be restarted
+ * any time [workflow], [diagnosticListener], or the `CoroutineContext`
+ * changes. The runtime will be cancelled when this function stops composing.
+ *
+ * @param workflow The [Workflow] to render.
+ * @param viewEnvironment The [ViewEnvironment] used to show the [ComposeRendering]s emitted by
+ * the workflow.
+ * @param props The props to render the root workflow with. If this value changes between calls,
+ * the workflow runtime will re-render with the new props.
+ * @param onOutput A function that will be invoked any time the root workflow emits an output.
+ * @param diagnosticListener A [WorkflowDiagnosticListener] to configure on the runtime.
+ */
+@Composable fun <PropsT, OutputT : Any> WorkflowContainer(
+  workflow: Workflow<PropsT, OutputT, ComposeRendering>,
+  viewEnvironment: ViewEnvironment,
+  props: PropsT,
+  onOutput: (OutputT) -> Unit,
+  modifier: Modifier = Modifier,
+  diagnosticListener: WorkflowDiagnosticListener? = null
+) {
+  WorkflowContainer(workflow, props, onOutput, modifier, diagnosticListener) { rendering ->
+    rendering.render(viewEnvironment)
```

Comment:
This isn't applying the `ComposeViewFactoryRoot`. Since we're already in Compose land, it's not necessary to use that tool, but we should still handle that case correctly. Needs a test as well.

Reply:
Also, this should be applied in the non-ComposeWorkflow case as well. That's going to require changing that function to accept a `ViewEnvironment`, which is too bad